File: src/bot/bot.ts
import { 
  Client, 
  GatewayIntentBits, 
  ActivityType, 
  ClientOptions,
  Collection,
  REST,
  Routes
} from 'discord.js';
import { CommandTree } from './command-tree';
import { DB } from '../db/mongodb';
import { getRepoVersion, getNow } from '../static/utils';
import { Config } from '../db/structures';
import { AppCommandTranslator } from '../l10n';
import { readdirSync } from 'fs';
import { join } from 'path';

export class Zenox extends Client {
  owner_id: number;
  guild_id: number;
  uptime: Date;
  repo: any;
  version: string;
  env: string;
  config: Config;
  db: typeof DB;
  logWebhookUrl: string | undefined;
  commands: Collection<string, any>;

  constructor(env: string, config: Config) {
    const options: ClientOptions = {
      intents: [
        GatewayIntentBits.Guilds,
        GatewayIntentBits.GuildMessages,
        GatewayIntentBits.MessageContent,
        GatewayIntentBits.GuildMembers
      ],
      presence: {
        activities: [{
          name: `${getRepoVersion() || 'dev'} | ZZZ Full Support`,
          type: ActivityType.Custom
        }]
      }
    };

    super(options);

    this.owner_id = 585834029484343298;
    this.guild_id = 1129777497454686330;
    this.uptime = getNow();
    this.repo = null; // Will be set if git is available
    this.version = getRepoVersion() || "dev";
    this.env = env;
    this.config = config;
    this.db = DB;
    this.logWebhookUrl = process.env[`LOGS_WEBHOOK_${env.toUpperCase()}`];
    this.commands = new Collection();

    this.setupEventHandlers();
  }

  private setupEventHandlers(): void {
    this.on('ready', () => {
      console.log(`Logged in as ${this.user?.tag}`);
    });

    this.on('error', (error) => {
      console.error('Bot error:', error);
      this.captureException(error);
    });
  }

  async setupHook(): Promise<void> {
    // Set Translator
    // Note: Discord.js v14 doesn't have built-in translator like discord.py
    // You might need to implement your own translation system

    // Load Cogs
    const cogsPath = join(__dirname, '../cogs');
    try {
      const cogFiles = readdirSync(cogsPath).filter(file => file.endsWith('.js'));
      
      for (const file of cogFiles) {
        const cogName = file.replace('.js', '');
        try {
          const cog = require(join(cogsPath, file));
          if (cog.default) {
            const cogInstance = new cog.default(this);
            this.commands.set(cogInstance.name, cogInstance);
          }
        } catch (error) {
          console.error(`Failed to load cog ${cogName}:`, error);
          this.captureException(error as Error);
        }
      }
    } catch (error) {
      console.error('Error loading cogs:', error);
    }

    // Register commands
    await this.registerCommands();
  }

  private async registerCommands(): Promise<void> {
    if (!this.user) return;

    const rest = new REST({ version: '10' }).setToken(this.token!);
    const commands = Array.from(this.commands.values()).map(cmd => cmd.data);

    try {
      console.log('Started refreshing application (/) commands.');

      await rest.put(
        Routes.applicationCommands(this.user.id),
        { body: commands }
      );

      console.log('Successfully reloaded application (/) commands.');
    } catch (error) {
      console.error('Error registering commands:', error);
    }
  }

  captureException(e: Error): void {
    // Implement Sentry or other error tracking here
    console.error('Captured exception:', e);
    
    // Filter out specific errors if needed
    if (e.message.includes('10062')) {
      return;
    }
    
    // You can integrate with Sentry here
    // Sentry.captureException(e);
  }
}

File: zenox/bot/bot.py
import discord
import sentry_sdk
import os
import git
from discord.ext import commands
from aiohttp import ClientSession
from loguru import logger

from pathlib import Path

from.command_tree import CommandTree
from ..db.mongodb import DB
from..static.utils import get_repo_version, get_now
from ..db.structures import Config
from ..l10n import Translator, AppCommandTranslator

class Zenox(commands.AutoShardedBot):
    owner_id: int

    def __init__(
        self,
        *,
        env: str,
        config: Config
    ) -> None:
        self.owner_id = 585834029484343298
        self.guild_id = 1129777497454686330
        self.uptime = get_now()
        self.repo = git.Repo()
        self.version = get_repo_version() or "dev"
        self.env = env
        self.config = config
        self.db = DB
        self.log_webhook_url = os.getenv(f"LOGS_WEBHOOK_{env.upper()}")

        super().__init__(
            command_prefix=commands.when_mentioned,
            intents=discord.Intents.default(),
            case_insensitive=True,
            help_command=None,
            tree_cls=CommandTree,
            allowed_contexts=discord.app_commands.AppCommandContext(guild=True, dm_channel=True, private_channel=False),
            allowed_installs=discord.app_commands.AppInstallationType(guild=True, user=False),
            activity=discord.CustomActivity(f"{self.version} | ZZZ Full Support")
        )

    async def setup_hook(self) -> None:
        # Set Translator
        await self.tree.set_translator(AppCommandTranslator())

        # Load Cogs
        for filepath in Path("zenox/cogs").glob("*.py"):
            cog_name = Path(filepath).stem
            try:
                await self.load_extension(f"zenox.cogs.{cog_name}")
            except Exception as e:
                print(f"Failed to load cog {cog_name!r}")
                self.capture_exception(e)
        return await super().setup_hook()
    
    def capture_exception(self, e: Exception) -> None:
        if isinstance(e, discord.NotFound) and e.code == 10062:
            return
        sentry_sdk.capture_exception(e)